synapse/federation/federation_server.py

```diff
@@ -949,7 +949,10 @@ async def _process_incoming_pdus_in_room_inner(
                 room_id, room_version
             )
             if not next:
-                return
+                # We need to ensure that we always use the lock, so that it gets
+                # correctly dropped.
+                async with lock:
+                    return
```

Comment:
I can't help wondering if `Lock` should really be implementing the context manager interface. What's the advantage over just mandating that people call `lock.release()` ?

Reply:
Personally I find using a context manager interface clearer than using a `try/finally`. We could also have an explicit `lock.release()` to use here

Reply:
> Personally I find using a context manager interface clearer than using a try/finally

well maybe, but that seems a bit irrelevant.

basically:

```
lock.release()
```

seems a more intuitive way of dropping a lock than 

```
async with lock():
    pass
```

I think it's basically a bit silly to have an `aenter` which does nothing.

Reply:
I've added a `lock.release()`  method



> > Personally I find using a context manager interface clearer than using a try/finally
> 
> well maybe, but that seems a bit irrelevant.

How so? The other place we use the `async with lock` we'd need to convert to a `try/finally` if we got rid of the context manager.

Reply:
> How so? The other place we use the `async with lock` we'd need to convert to a `try/finally` if we got rid of the context manager.

Oh I see. Well, yes, we would, and tbh I think that would be an improvement. Currently it looks like something magical happens at the start of that `async with` block, which is misleading. The `with` mechanism doesn't really work very well here, because in each case we already hold the resource (ie, the lock) when we enter the `with` block.

Instead we could put the whole body of `_process_incoming_pdus_in_room_inner` inside a `try` block:

```python
async def _process_incoming_pdus_in_room_inner(...):
    try:
        # ...
        while True:
            await self.handler.on_receive_pdu(event)

            # drop the lock before checking for more events
            await lock.release()
            lock = None

            next = await self.store.get_next_staged_event_for_room()
            if not next:
                break
            
            lock = await self.store.try_acquire_lock()
            if not lock:
                break
    finally:
        if lock:
            lock.release()
```

... thus ensuring that we always drop the lock even if we end up bailing out at odd points due to exceptions or whatever.


Anyway, what you have now is fine.